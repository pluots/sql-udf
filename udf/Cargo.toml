[package]
name = "udf"
version = "0.1.1"
rust-version = "1.65"
edition = "2021"
description = "Easily create user defined functions (UDFs) for MariaDB and MySQL."
repository = "https://github.com/pluots/sql-udf"
readme = "../README.md"
license = "Apache-2.0"
keywords = ["sql", "udf"]

[dependencies]
mysqlclient-sys = "0.2.5"
udf-macros = { path = "../udf_macros", version = "0.1.0"}

[dev-dependencies]
lipsum = "0.8.2"
# Needed for 
uuid = { version = "1.1.2", features = ["v1", "v3", "v4", "v5", "fast-rng"] }

[[example]]
name = "uuidv"
crate-type = ["cdylib"]

[[example]]
name = "basic_sum"
crate-type = ["cdylib"]

[[example]]
name = "lipsum"
crate-type = ["cdylib"]

[[example]]
name = "sequence"
crate-type = ["cdylib"]

[[example]]
name = "avgcost"
crate-type = ["cdylib"]

[[example]]
name = "is_const"
crate-type = ["cdylib"]

[[example]]
name = "median"
crate-type = ["cdylib"]

[[example]]
name = "metaphon"
crate-type = ["cdylib"]


[package.metadata.release]
shared-version = true

# Can't run replacements at workspace root. Need to use this "hacky" sort of way.
# [[package.metadata.release.pre-release-replacements]]
# file = "../../CHANGELOG.md"
# search = "Unreleased"
# replace = "{{version}}"

# [[package.metadata.release.pre-release-replacements]]
# file = "../../CHANGELOG.md"
# search = "\\.\\.\\.HEAD"
# replace = "...{{tag_name}}"
# exactly = 1

# [[package.metadata.release.pre-release-replacements]]
# file = "../../CHANGELOG.md"
# search = "ReleaseDate"
# replace = "{{date}}"

# [[package.metadata.release.pre-release-replacements]]
# file = "../../CHANGELOG.md"
# search = "<!-- next-header -->"
# replace = """\
#   <!-- next-header -->\n\n\
#   ## [Unreleased] - ReleaseDate\n\n\
#   ### Added\n\n\
#   ### Changed\n\n\
#   ### Removed\n\n\
#   """
# exactly = 1

# [[package.metadata.release.pre-release-replacements]]
# file = "../../CHANGELOG.md"
# search = "<!-- next-url -->"
# replace = """\
#   <!-- next-url -->\n\
#   [Unreleased]: https://github.com/pluots/stringmetrics/compare/{{tag_name}}...HEAD\
#   """
# exactly = 1
