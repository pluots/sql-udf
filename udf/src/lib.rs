//! A wrapper crate to make writing SQL user-defined functions (UDFs) easy
//!
//! This crate provides bindings for easy creation of SQL user-defined functions
//! in Rust. See [the
//! readme](https://github.com/pluots/sql-udf/blob/main/README.md) for more
//! background information.
//!
//! # Usage
//!
//! Using this crate is fairly simple: create a struct that will be used to
//! share data among UDF functions (which can be zero-sized), then implement
//! needed traits for it. [`BasicUdf`] provides function signatures for standard
//! UDFs, and [`AggregateUdf`] provides signatures for aggregate (and window)
//! UDFs. See the documentation there for a step-by-step guide.
//!
//! ```
//! use udf::prelude::*;
//!
//! // Our struct that will produce a UDF of name `my_udf`
//! struct MyUdf {}
//!
//! #[register]
//! impl BasicUdf for MyUdf {
//!     // Specify return type of this UDF to be a nullable integer
//!     type Returns<'a> = Option<i64>;
//!
//!     // Perform initialization steps here
//!     fn init<'a>(
//!         cfg: &UdfCfg<Init>,
//!         args: &'a ArgList<'a, Init>
//!     ) -> Result<Self, String> {
//!         todo!();
//!     }
//!
//!     // Create a result here
//!     fn process<'a>(
//!         &'a mut self,
//!         cfg: &UdfCfg<Process>,
//!         args: &ArgList<Process>,
//!         error: Option<NonZeroU8>,
//!     ) -> Result<Self::Returns<'a>, ProcessError> {
//!         todo!();
//!     }
//! }
//! ```
//!
//! # Building & Usage
//!
//! The above example will create three C-callable functions: `my_udf`,
//! `my_udf_init`, and `my_udf_deinit`, which is what `MariaDB` and `MySql`
//! expect for UDFs. To create a C dynamic library (as is required for usage),
//! add the following to your `Cargo.toml`
//!
//! ```toml
//! [lib]
//! crate-type = ["cdylib"]
//! ```
//!
//! The next time you run `cargo build --release`, in `target/release` there
//! will be a shared library `.so` file. Copy this to your `plugin_dir` location
//! (usually `/usr/lib/mysql/plugin/`), and load the function with the
//! following:
//!
//! ```sql
//! CREATE FUNCTION my_udf RETURNS integer SONAME 'libudf_test.so';
//! ```
//!
//! Replace `my_udf` with the function name, `integer` with the return type, and
//! `libudf_test.so` with the correct file name.
//!
//! More details on building are discussed in [the project
//! readme](https://github.com/pluots/sql-udf/blob/main/README.md).
//!
//! # Version Note
//!
//! Because of reliance on a feature called GATs, this library requires Rust
//! version >= 1.65. This only became stable on 2022-11-03; if you encounter
//! issues compiling, be sure to update your toolchain.
//!
//! ```sh
//! rustup update
//! ```

// Strict clippy
#![warn(
    clippy::pedantic,
    // clippy::cargo,
    clippy::nursery,
    clippy::str_to_string,
    clippy::missing_inline_in_public_items,
    clippy::exhaustive_enums,
    clippy::pattern_type_mismatch
)]
// Pedantic config
#![allow(
    clippy::missing_const_for_fn,
    clippy::missing_panics_doc,
    clippy::must_use_candidate,
    clippy::cast_possible_truncation
)]

// We re-export this so we can use it in our macro, but don't need it
// to show up in our docs
#[doc(hidden)]
pub extern crate chrono;

extern crate udf_macros;
pub extern crate udf_sys;

pub use udf_macros::register;

pub mod prelude;
pub mod traits;
pub mod types;

// We hide this because it's really only used by our proc macros
#[doc(hidden)]
pub mod wrapper;

#[doc(inline)]
pub use traits::*;
#[doc(inline)]
pub use types::{MYSQL_ERRMSG_SIZE, *};

/// Print a formatted log message to `stderr` to display in server logs
///
/// Performs formatting to match other common SQL error logs, roughly:
///
/// ```text
/// 2022-10-15 13:12:54+00:00 [Warning] Udf: this is the message
/// ```
///
/// ```
/// use udf::udf_log;
///
/// // Prints "2022-10-08 05:27:30+00:00 [Error] UDF: this is an error"
/// // This matches the default entrypoint log format
/// udf_log!(Error: "this is an error");
///
/// udf_log!(Warning: "this is a warning");
///
/// udf_log!(Note: "this is info");
///
/// udf_log!(Debug: "this is a debug message");
///
/// udf_log!("i print without the '[Level] UDF:' formatting");
/// ```
#[macro_export]
macro_rules! udf_log {
    (Error: $msg:expr) => {
        let formatted = format!("[Error] UDF: {}", $msg);
        udf_log!(formatted);
    };
    (Warning: $msg:expr) => {
        let formatted = format!("[Warning] UDF: {}", $msg);
        udf_log!(formatted);
    };
    (Note: $msg:expr) => {
        let formatted = format!("[Note] UDF: {}", $msg);
        udf_log!(formatted);
    };
    (Debug: $msg:expr) => {
        let formatted = format!("[Debug] UDF: {}", $msg);
        udf_log!(formatted);
    };
    ($msg:tt) => {
        eprintln!(
            "{} {}",
            udf::chrono::Utc::now().format("%Y-%m-%d %H:%M:%S%:z"),
            $msg
        );
    };
}
