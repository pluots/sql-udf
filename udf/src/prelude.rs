//! Module that can be imported with `use udf::prelude::*;` to quickly get the
//! most often used imports.

pub use std::num::NonZeroU8;

pub use crate::register;
pub use crate::traits::{AggregateUdf, BasicUdf};
pub use crate::types::{ArgList, Init, InitCfg, Process, ProcessError, SqlArg, SqlType};
