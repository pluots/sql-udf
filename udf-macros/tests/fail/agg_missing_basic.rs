#![allow(unused)]

use udf::prelude::*;

struct MyUdf;

impl AggregateUdf for MyUdf {
    // Required methods
    fn clear(&mut self, cfg: &UdfCfg<Process>, error: Option<NonZeroU8>) -> Result<(), NonZeroU8> {
        todo!()
    }
    fn add(
        &mut self,
        cfg: &UdfCfg<Process>,
        args: &ArgList<'_, Process>,
        error: Option<NonZeroU8>,
    ) -> Result<(), NonZeroU8> {
        todo!()
    }
}

fn main() {}
