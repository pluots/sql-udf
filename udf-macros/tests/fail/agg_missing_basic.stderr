error[E0277]: the trait bound `MyUdf: BasicUdf` is not satisfied
 --> tests/fail/agg_missing_basic.rs:7:23
  |
7 | impl AggregateUdf for MyUdf {
  |                       ^^^^^ the trait `BasicUdf` is not implemented for `MyUdf`
  |
note: required by a bound in `udf::AggregateUdf`
 --> $WORKSPACE/udf/src/traits.rs
  |
  | pub trait AggregateUdf: BasicUdf {
  |                         ^^^^^^^^ required by this bound in `AggregateUdf`
